src/transport/NetworkProvisioning.cpp

```diff
@@ -229,6 +229,44 @@ CHIP_ERROR NetworkProvisioning::SendNetworkCredentials(const char * ssid, const
     return err;
 }
 
+CHIP_ERROR NetworkProvisioning::SendThreadCredentials(const DeviceLayer::Internal::DeviceNetworkInfo & threadData)
+{
+    CHIP_ERROR err                = CHIP_NO_ERROR;
+    System::PacketBuffer * buffer = System::PacketBuffer::New();
+
+    ChipLogProgress(NetworkProvisioning, "Sending Thread Credentials");
+    VerifyOrExit(mDelegate != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(buffer != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    {
+        BufBound bbuf(buffer->Start(), buffer->AvailableDataLength());
+        bbuf.Put(threadData.ThreadNetworkName, sizeof(threadData.ThreadNetworkName));
+        bbuf.Put(threadData.ThreadExtendedPANId, sizeof(threadData.ThreadExtendedPANId));
+        bbuf.Put(threadData.ThreadMeshPrefix, sizeof(threadData.ThreadMeshPrefix));
+        bbuf.Put(threadData.ThreadMasterKey, sizeof(threadData.ThreadMasterKey));
+        bbuf.Put(threadData.ThreadPSKc, sizeof(threadData.ThreadPSKc));
+        bbuf.PutLE16(threadData.ThreadPANId);
+        bbuf.Put(threadData.ThreadChannel);
+        bbuf.Put(static_cast<uint8_t>(threadData.FieldPresent.ThreadExtendedPANId));
+        bbuf.Put(static_cast<uint8_t>(threadData.FieldPresent.ThreadMeshPrefix));
+        bbuf.Put(static_cast<uint8_t>(threadData.FieldPresent.ThreadPSKc));
+
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        buffer->SetDataLength(static_cast<uint16_t>(bbuf.Written()));
+
+        err = mDelegate->SendSecureMessage(Protocols::kChipProtocol_NetworkProvisioning,
```

Comment:
Other SendMessage implementations usually ::Free the buffer regardless of error code. Could you double check if this is the case or not? (i.e. is this ok or should we prevent freeing if sendmessage is called).

Reply:
I copied this pattern from other methods in NetworkProvisioning, but after re-checking you're right. SendSecureMessage behaves inconsistently in this matter:
1) if it fails in its own body it doesn't free the buffer
2) if it reaches mTransport->SendMessage() the buffer is freed regardless of the outcome
Which means that if mTransport->SendMessage() fails we may release a single buffer twice. How about I log an issue for that and fix it in a separate PR?

Reply:
#3393 logged